# Banking Intent — Local Training & Evaluation

This notebook is configured to run on your local machine using a `.env` file for credentials.

In [1]:
import os
from dotenv import load_dotenv

# Load environment variables from the .env file located one level up
load_dotenv(".env")

# Set variables for Hugging Face and LangSmith
os.environ["HF_TOKEN"] = os.getenv("HF_TOKEN", "")
os.environ["LANGCHAIN_API_KEY"] = os.getenv("YOUR_LANGSMITH_API_KEY", "")
os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_PROJECT"] = "banking-intent-unsloth"
print("Environment variables loaded from .env")

Environment variables loaded from .env


In [2]:
# === CONFIG ===
HF_REPO_ID = "TQZinh/banking-intent-unsloth"
WORKING_DIR = os.getcwd()
CHECKPOINT_DIR = os.path.join(WORKING_DIR, "outputs", "checkpoint")
os.makedirs(CHECKPOINT_DIR, exist_ok=True)

print(f"Working Directory: {WORKING_DIR}")
print(f"Checkpoint Directory: {CHECKPOINT_DIR}")

Working Directory: c:\Users\VINH\OneDrive - VNU-HCMUS\Attachments\Desktop\YEAR 3\ƯDNLP\lab_2\banking-intent-unsloth
Checkpoint Directory: c:\Users\VINH\OneDrive - VNU-HCMUS\Attachments\Desktop\YEAR 3\ƯDNLP\lab_2\banking-intent-unsloth\outputs\checkpoint


## 1. Prepare data

In [3]:
os.chdir(os.path.join(WORKING_DIR, "scripts"))
!python preprocess_data.py
print("Data preprocessing complete.")

Loading BANKING77 dataset from Hugging Face (parquet)...
Creating representative training subset (5,000 samples)...
Final Train size: 5000
Final Test size:  3080
Saved train.csv and test.csv to the sample_data directory.
Data preprocessing complete.


## 2. Configure HF repo for checkpoint push

In [4]:
import yaml

config_path = f"{WORKING_DIR}/configs/train.yaml"
with open(config_path, 'r', encoding='utf-8') as f:
    config = yaml.safe_load(f)

config['hub_model_id'] = HF_REPO_ID
config['output_dir'] = CHECKPOINT_DIR

with open(config_path, 'w', encoding='utf-8') as f:
    yaml.dump(config, f, default_flow_style=False, allow_unicode=True)

print(f"hub_model_id  : {HF_REPO_ID}")
print(f"output_dir    : {CHECKPOINT_DIR}")


hub_model_id  : TQZinh/banking-intent-unsloth
output_dir    : c:\Users\VINH\OneDrive - VNU-HCMUS\Attachments\Desktop\YEAR 3\ƯDNLP\lab_2\banking-intent-unsloth\outputs\checkpoint


## 3. Train

If this cell is re-run after a session restart, it will **automatically resume** from the latest local checkpoint (if any), or pull from Hub if disk was wiped.

In [5]:
import glob
from huggingface_hub import snapshot_download

output_dir = f"{WORKING_DIR}/outputs/checkpoint"
os.makedirs(output_dir, exist_ok=True)

has_local_checkpoint = bool(glob.glob(os.path.join(output_dir, "checkpoint-*")))

if not has_local_checkpoint:
    try:
        print(f"No local checkpoint. Trying to restore from Hub: {HF_REPO_ID}")
        snapshot_download(
            repo_id=HF_REPO_ID,
            local_dir=output_dir,
            token=os.environ["HF_TOKEN"],
        )
        print("Restored from Hub.")
    except Exception as e:
        print(f"Hub restore skipped ({e}) — starting fresh.")
else:
    latest = sorted(glob.glob(os.path.join(output_dir, "checkpoint-*")))[-1]
    print(f"Local checkpoint found: {latest}")

No local checkpoint. Trying to restore from Hub: TQZinh/banking-intent-unsloth


c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Fetching 2 files: 100%|██████████| 2/2 [00:00<00:00, 95.75it/s]

Restored from Hub.


In [6]:
os.chdir(os.path.join(WORKING_DIR, "scripts"))
%run train.py
print("Training process finished.")

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


W0427 11:08:41.589000 22300 site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.


🦥 Unsloth Zoo will now patch everything to make training faster!
Logged in to HuggingFace Hub.
No checkpoint found — starting fresh.
Loading model: unsloth/Qwen3-4B-unsloth-bnb-4bit
==((====))==  Unsloth 2026.4.6: Fast Qwen3 patching. Transformers: 4.57.0.
   \\   /|    NVIDIA GeForce RTX 3070 Laptop GPU. Num GPUs = 1. Max memory: 8.0 GB. Platform: Windows.
O^O/ \_/ \    Torch: 2.7.1+cu118. CUDA: 8.6. CUDA Toolkit: 11.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
unsloth/Qwen3-4B-unsloth-bnb-4bit does not have a padding token! Will use pad_token = <|PAD_TOKEN|>.


Unsloth: Dropout = 0 is supported for fast patching. You are using dropout = 0.05.
Unsloth will patch all other layers, except LoRA matrices, causing a performance hit.
Unsloth 2026.4.6 patched 36 layers with 0 QKV layers, 0 O layers and 0 MLP layers.


Loading data from ../sample_data/train.csv...


Map: 100%|██████████| 5000/5000 [00:00<00:00, 7523.40 examples/s]
Unsloth: Tokenizing ["formatted_text"] (num_proc=2): 100%|██████████| 5000/5000 [00:24<00:00, 203.45 examples/s]


🦥 Unsloth: Padding-free auto-enabled, enabling faster training.
Training...


==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 5,000 | Num Epochs = 1 | Total steps = 625
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 33,030,144 of 4,055,498,240 (0.81% trained)


Unsloth: Will smartly offload gradients to save VRAM!


Step,Training Loss
10,2.254300
20,0.565300
30,0.111100
40,0.096400
50,0.070600
60,0.074900
70,0.065700
80,0.079500
90,0.073600
100,0.068300


'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: b3a66a65-714e-4a66-a230-421ad58da861)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
[huggingface_hub.utils._http|WARNING]'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: b3a66a65-714e-4a66-a230-421ad58da861)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
Retrying in 1s [Retry 1/5].
[huggingface_hub.utils._http|WARNING]Retrying in 1s [Retry 1/5].
'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: 77ed53b8-4b46-4a7d-9383-94b0258d8a26)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
[huggingface_hub.utils._http|WARNING]'(Protoc


[HubPush] step 250 → pushing to TQZinh/banking-intent-unsloth


Processing Files (1 / 1): 100%|██████████|  132MB /  132MB, 6.09MB/s  
New Data Upload: 100%|██████████|  132MB /  132MB, 6.09MB/s  


Saved model to https://huggingface.co/TQZinh/banking-intent-unsloth


Processing Files (1 / 1): 100%|██████████| 11.4MB / 11.4MB, 8.13MB/s  
New Data Upload: |          |  0.00B /  0.00B,  0.00B/s  
'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: 29a2ba02-3a99-414a-a442-06e2bd460756)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
[huggingface_hub.utils._http|WARNING]'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: 29a2ba02-3a99-414a-a442-06e2bd460756)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
Retrying in 1s [Retry 1/5].
[huggingface_hub.utils._http|WARNING]Retrying in 1s [Retry 1/5].
'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: 1c115cd6-c742-4d78-871f-c51fe9188bd8)')' thrown while requesting HEA


[HubPush] step 500 → pushing to TQZinh/banking-intent-unsloth


Processing Files (1 / 1): 100%|██████████|  132MB /  132MB, 6.31MB/s  
New Data Upload: 100%|██████████|  132MB /  132MB, 6.31MB/s  


Saved model to https://huggingface.co/TQZinh/banking-intent-unsloth


Processing Files (1 / 1): 100%|██████████| 11.4MB / 11.4MB,  0.00B/s  
New Data Upload: |          |  0.00B /  0.00B,  0.00B/s  
No files have been modified since last commit. Skipping to prevent empty commit.
[huggingface_hub.hf_api|WARNING]No files have been modified since last commit. Skipping to prevent empty commit.
'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: b3438ad1-6239-497c-ae15-b9e372713181)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
[huggingface_hub.utils._http|WARNING]'(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')), '(Request ID: b3438ad1-6239-497c-ae15-b9e372713181)')' thrown while requesting HEAD https://huggingface.co/unsloth/Qwen3-4B-unsloth-bnb-4bit/resolve/main/config.json
Retrying in 1s [Retry 1/5].
[huggingface_hub.utils._http|WARNING]Retrying in 1s [Retry 1/5

Saving final adapter to c:\Users\VINH\OneDrive - VNU-HCMUS\Attachments\Desktop\YEAR 3\ƯDNLP\lab_2\banking-intent-unsloth\outputs\checkpoint...
Pushing final adapter to Hub: TQZinh/banking-intent-unsloth


Processing Files (1 / 1): 100%|██████████|  132MB /  132MB, 6.37MB/s  
New Data Upload: 100%|██████████|  132MB /  132MB, 6.37MB/s  


Saved model to https://huggingface.co/TQZinh/banking-intent-unsloth


Processing Files (1 / 1): 100%|██████████| 11.4MB / 11.4MB,  0.00B/s  
New Data Upload: |          |  0.00B /  0.00B,  0.00B/s  


Done.
Training process finished.


## 4. Quick sanity check — predict one sample

In [3]:
import sys
import yaml
import os

sys.path.insert(0, f"{WORKING_DIR}/scripts")

# Tương tác với inference.yaml
infer_config_path = f"{WORKING_DIR}/configs/inference.yaml"
with open(infer_config_path, 'r', encoding='utf-8') as f:
    infer_config = yaml.safe_load(f)

infer_config['langsmith_api_key'] = os.environ.get("LANGCHAIN_API_KEY", "")
infer_config['langsmith_project'] = os.environ.get("LANGCHAIN_PROJECT", "banking-intent-unsloth")
infer_config['model_path'] = CHECKPOINT_DIR

with open(infer_config_path, 'w', encoding='utf-8') as f:
    yaml.dump(infer_config, f, default_flow_style=False, allow_unicode=True)

print("Updated inference.yaml with LangSmith credentials and model path.\n")

from inference import IntentClassification

clf = IntentClassification(infer_config_path, mode="finetuned")
test_input = "I am still waiting on my card?"
result = clf.predict(test_input)
print(result)
print("The right one: card_arrival")


Updated inference.yaml with LangSmith credentials and model path.

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
W0427 21:43:38.691000 25364 site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.


🦥 Unsloth Zoo will now patch everything to make training faster!
[FINETUNED] Loading model: C:\Users\VINH\OneDrive - VNU-HCMUS\Attachments\Desktop\YEAR 3\ƯDNLP\lab_2\banking-intent-unsloth\outputs\checkpoint
==((====))==  Unsloth 2026.4.6: Fast Qwen3 patching. Transformers: 4.57.0.
   \\   /|    NVIDIA GeForce RTX 3070 Laptop GPU. Num GPUs = 1. Max memory: 8.0 GB. Platform: Windows.
O^O/ \_/ \    Torch: 2.7.1+cu118. CUDA: 8.6. CUDA Toolkit: 11.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Unsloth 2026.4.6 patched 36 layers with 0 QKV layers, 0 O layers and 0 MLP layers.


LangSmith tracing enabled — project: banking-intent-unsloth
{'input': 'I am still waiting on my card?', 'raw_output': 'Card arrival', 'label': 'card_arrival'}
The right one: card_arrival


In [5]:
import gc
import torch

# Xoá model sanity check cũ nếu đang tồn tại
if 'clf' in globals():
    del clf
    
gc.collect()
torch.cuda.empty_cache()
import sys, gc, yaml
import torch
import pandas as pd
from sklearn.metrics import accuracy_score, classification_report
from tqdm import tqdm

sys.path.insert(0, f"{WORKING_DIR}/scripts")
from inference import IntentClassification

with open(f"{WORKING_DIR}/configs/inference.yaml") as f:
    config = yaml.safe_load(f)
batch_size = config.get("batch_size", 8)

test_df = pd.read_csv(f"{WORKING_DIR}/sample_data/test.csv")

# Stratified 200 samples: 3 per intent (77*3=231) then subsample to 200
sample_df = (
    test_df.groupby("intent_name", group_keys=False)
    .sample(n=3, random_state=42)
    .sample(n=200, random_state=42)
    .reset_index(drop=True)
)
print(f"Sampled {len(sample_df)} rows across {sample_df['intent_name'].nunique()} intents")


def run_eval(clf, df, label, print_per_sample=True):
    texts = df["text"].tolist()
    y_true = df["intent_name"].tolist()
    y_pred = []
    for start in tqdm(range(0, len(texts), batch_size), desc=label):
        results = clf.predict_batch(texts[start:start + batch_size])
        for result, true_label in zip(results, y_true[start:start + batch_size]):
            y_pred.append(result["label"])
            if print_per_sample:
                status = "OK" if result["label"] == true_label else "MISS"
                print(f"  [{status}] true={true_label}  pred={result['label']}", flush=True)
    acc = accuracy_score(y_true, y_pred)
    print(f"\n>>> {label} accuracy: {acc:.4f} ({int(acc*len(df))}/{len(df)})\n", flush=True)
    print(classification_report(y_true, y_pred, digits=4), flush=True)
    return acc


def free_model(clf):
    del clf.model
    del clf.tokenizer
    del clf
    gc.collect()
    torch.cuda.empty_cache()
    print(f"VRAM freed: {torch.cuda.memory_allocated()/1e9:.2f}GB used", flush=True)


🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
W0427 16:44:52.925000 27036 site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.


🦥 Unsloth Zoo will now patch everything to make training faster!
Sampled 200 rows across 77 intents


## 5. Evaluate — quick test (200 samples)

Stratified sample across intents to verify the pipeline before running the full evaluation.

In [5]:
print("=" * 60)
print("EVALUATE — 200 samples")
print("=" * 60)

results = {}

clf = IntentClassification(f"{WORKING_DIR}/configs/inference.yaml", mode="zero_shot")
results["zero_shot"] = run_eval(clf, sample_df, "zero_shot")
free_model(clf)

clf = IntentClassification(f"{WORKING_DIR}/configs/inference.yaml", mode="finetuned")
results["finetuned"] = run_eval(clf, sample_df, "finetuned")
free_model(clf)

print("=== SUMMARY ===")
for mode, acc in results.items():
    print(f"  {mode:<12} {acc:.4f}  ({acc*100:.2f}%)")

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
W0427 14:22:40.684000 2480 site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.


🦥 Unsloth Zoo will now patch everything to make training faster!
Sampled 200 rows across 77 intents
EVALUATE — 200 samples
[ZERO_SHOT] Loading model: unsloth/Qwen3-4B-unsloth-bnb-4bit
==((====))==  Unsloth 2026.4.6: Fast Qwen3 patching. Transformers: 4.57.0.
   \\   /|    NVIDIA GeForce RTX 3070 Laptop GPU. Num GPUs = 1. Max memory: 8.0 GB. Platform: Windows.
O^O/ \_/ \    Torch: 2.7.1+cu118. CUDA: 8.6. CUDA Toolkit: 11.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
unsloth/Qwen3-4B-unsloth-bnb-4bit does not have a padding token! Will use pad_token = <|PAD_TOKEN|>.
LangSmith tracing enabled — project: banking-intent-unsloth


zero_shot:   0%|          | 0/25 [00:00<?, ?it/s]

  [OK] true=virtual_card_not_working  pred=virtual_card_not_working
  [OK] true=change_pin  pred=change_pin
  [OK] true=apple_pay_or_google_pay  pred=apple_pay_or_google_pay
  [MISS] true=top_up_by_bank_transfer_charge  pred=transfer_fee_charged
  [OK] true=automatic_top_up  pred=automatic_top_up
  [OK] true=transfer_not_received_by_recipient  pred=transfer_not_received_by_recipient
  [MISS] true=beneficiary_not_allowed  pred=transfer_not_received_by_recipient
  [OK] true=transfer_into_account  pred=transfer_into_account


zero_shot:   4%|▍         | 1/25 [00:51<20:42, 51.78s/it]

  [OK] true=receiving_money  pred=receiving_money
  [OK] true=lost_or_stolen_card  pred=lost_or_stolen_card
  [OK] true=verify_top_up  pred=verify_top_up
  [MISS] true=balance_not_updated_after_bank_transfer  pred=pending_transfer
  [OK] true=exchange_charge  pred=exchange_charge
  [OK] true=top_up_failed  pred=top_up_failed
  [OK] true=top_up_by_cash_or_cheque  pred=top_up_by_cash_or_cheque
  [OK] true=passcode_forgotten  pred=passcode_forgotten


zero_shot:   8%|▊         | 2/25 [02:30<30:28, 79.52s/it]

  [MISS] true=pending_top_up  pred=top_up_failed
  [OK] true=card_about_to_expire  pred=card_about_to_expire
  [OK] true=wrong_amount_of_cash_received  pred=wrong_amount_of_cash_received
  [MISS] true=top_up_reverted  pred=top_up_failed
  [OK] true=failed_transfer  pred=failed_transfer
  [OK] true=supported_cards_and_currencies  pred=supported_cards_and_currencies
  [MISS] true=unable_to_verify_identity  pred=verify_my_identity
  [OK] true=top_up_limits  pred=top_up_limits


zero_shot:  12%|█▏        | 3/25 [03:59<30:43, 83.78s/it]

  [OK] true=getting_virtual_card  pred=getting_virtual_card
  [MISS] true=balance_not_updated_after_bank_transfer  pred=pending_transfer
  [OK] true=pending_transfer  pred=pending_transfer
  [OK] true=exchange_rate  pred=exchange_rate
  [OK] true=exchange_via_app  pred=exchange_via_app
  [OK] true=declined_transfer  pred=declined_transfer
  [MISS] true=transfer_not_received_by_recipient  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=cash_withdrawal_charge  pred=cash_withdrawal_charge


zero_shot:  16%|█▌        | 4/25 [06:11<36:00, 102.88s/it]

  [MISS] true=get_physical_card  pred=passcode_forgotten
  [OK] true=card_not_working  pred=card_not_working
  [MISS] true=automatic_top_up  pred=lost_or_stolen_card
  [OK] true=lost_or_stolen_card  pred=lost_or_stolen_card
  [MISS] true=reverted_card_payment?  pred=declined_card_payment
  [MISS] true=get_physical_card  pred=change_pin
  [MISS] true=declined_cash_withdrawal  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=pending_top_up  pred=top_up_failed


zero_shot:  20%|██        | 5/25 [08:14<36:40, 110.04s/it]

  [OK] true=edit_personal_details  pred=edit_personal_details
  [MISS] true=transfer_fee_charged  pred=extra_charge_on_statement
  [OK] true=pending_cash_withdrawal  pred=pending_cash_withdrawal
  [OK] true=contactless_not_working  pred=contactless_not_working
  [MISS] true=fiat_currency_support  pred=supported_cards_and_currencies
  [MISS] true=fiat_currency_support  pred=supported_cards_and_currencies
  [MISS] true=pin_blocked  pred=change_pin
  [MISS] true=card_payment_wrong_exchange_rate  pred=wrong_exchange_rate_for_cash_withdrawal


zero_shot:  24%|██▍       | 6/25 [10:14<35:57, 113.57s/it]

  [MISS] true=wrong_exchange_rate_for_cash_withdrawal  pred=atm_support
  [MISS] true=compromised_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=top_up_by_bank_transfer_charge  pred=receiving_money
  [OK] true=failed_transfer  pred=failed_transfer
  [OK] true=getting_virtual_card  pred=getting_virtual_card
  [OK] true=declined_card_payment  pred=declined_card_payment
  [MISS] true=declined_transfer  pred=declined_card_payment
  [OK] true=edit_personal_details  pred=edit_personal_details


zero_shot:  28%|██▊       | 7/25 [12:16<34:48, 116.05s/it]

  [OK] true=visa_or_mastercard  pred=visa_or_mastercard
  [MISS] true=get_physical_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=card_arrival  pred=card_arrival
  [MISS] true=beneficiary_not_allowed  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=change_pin  pred=change_pin
  [MISS] true=topping_up_by_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=transfer_not_received_by_recipient  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=apple_pay_or_google_pay  pred=apple_pay_or_google_pay


zero_shot:  32%|███▏      | 8/25 [14:23<33:55, 119.76s/it]

  [OK] true=getting_spare_card  pred=getting_spare_card
  [OK] true=cancel_transfer  pred=cancel_transfer
  [OK] true=receiving_money  pred=receiving_money
  [OK] true=change_pin  pred=change_pin
  [OK] true=country_support  pred=country_support
  [OK] true=activate_my_card  pred=activate_my_card
  [MISS] true=card_payment_wrong_exchange_rate  pred=exchange_rate
  [OK] true=get_disposable_virtual_card  pred=get_disposable_virtual_card


zero_shot:  36%|███▌      | 9/25 [15:50<29:11, 109.44s/it]

  [OK] true=pending_card_payment  pred=pending_card_payment
  [MISS] true=transfer_fee_charged  pred=extra_charge_on_statement
  [MISS] true=cash_withdrawal_not_recognised  pred=compromised_card
  [OK] true=pending_cash_withdrawal  pred=pending_cash_withdrawal
  [OK] true=transfer_into_account  pred=transfer_into_account
  [MISS] true=passcode_forgotten  pred=change_pin
  [MISS] true=card_about_to_expire  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=atm_support  pred=atm_support


zero_shot:  40%|████      | 10/25 [17:56<28:36, 114.41s/it]

  [OK] true=card_acceptance  pred=card_acceptance
  [MISS] true=cancel_transfer  pred=transfer_not_received_by_recipient
  [MISS] true=card_linking  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=wrong_exchange_rate_for_cash_withdrawal  pred=exchange_rate
  [MISS] true=get_disposable_virtual_card  pred=disposable_card_limits
  [MISS] true=receiving_money  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=card_payment_not_recognised  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=order_physical_card  pred=order_physical_card


zero_shot:  44%|████▍     | 11/25 [20:02<27:31, 117.97s/it]

  [OK] true=top_up_reverted  pred=top_up_reverted
  [MISS] true=country_support  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=card_delivery_estimate  pred=card_arrival
  [OK] true=exchange_charge  pred=exchange_charge
  [MISS] true=pending_card_payment  pred=pending_transfer
  [OK] true=declined_card_payment  pred=declined_card_payment
  [OK] true=passcode_forgotten  pred=passcode_forgotten
  [MISS] true=beneficiary_not_allowed  pred=declined_transfer


zero_shot:  48%|████▊     | 12/25 [22:12<26:24, 121.87s/it]

  [OK] true=verify_source_of_funds  pred=verify_source_of_funds
  [MISS] true=top_up_by_card_charge  pred=exchange_charge
  [OK] true=request_refund  pred=request_refund
  [OK] true=Refund_not_showing_up  pred=Refund_not_showing_up
  [OK] true=Refund_not_showing_up  pred=Refund_not_showing_up
  [OK] true=country_support  pred=country_support
  [OK] true=card_not_working  pred=card_not_working
  [OK] true=exchange_rate  pred=exchange_rate


zero_shot:  52%|█████▏    | 13/25 [23:26<21:26, 107.17s/it]

  [MISS] true=fiat_currency_support  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=pending_card_payment  pred=pending_card_payment
  [OK] true=terminate_account  pred=terminate_account
  [OK] true=disposable_card_limits  pred=disposable_card_limits
  [MISS] true=declined_transfer  pred=declined_card_payment
  [MISS] true=reverted_card_payment?  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=pin_blocked  pred=pin_blocked
  [OK] true=exchange_charge  pred=exchange_charge


zero_shot:  56%|█████▌    | 14/25 [25:33<20:45, 113.22s/it]

  [OK] true=card_arrival  pred=card_arrival
  [OK] true=top_up_limits  pred=top_up_limits
  [OK] true=cash_withdrawal_charge  pred=cash_withdrawal_charge
  [MISS] true=balance_not_updated_after_cheque_or_cash_deposit  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=visa_or_mastercard  pred=visa_or_mastercard
  [OK] true=top_up_reverted  pred=top_up_reverted
  [OK] true=card_acceptance  pred=card_acceptance
  [MISS] true=apple_pay_or_google_pay  pred=wrong_exchange_rate_for_cash_withdrawal


zero_shot:  60%|██████    | 15/25 [27:36<19:20, 116.06s/it]

  [OK] true=wrong_amount_of_cash_received  pred=wrong_amount_of_cash_received
  [OK] true=top_up_limits  pred=top_up_limits
  [MISS] true=age_limit  pred=transfer_not_received_by_recipient
  [OK] true=cancel_transfer  pred=cancel_transfer
  [OK] true=pending_cash_withdrawal  pred=pending_cash_withdrawal
  [OK] true=visa_or_mastercard  pred=visa_or_mastercard
  [MISS] true=why_verify_identity  pred=verify_my_identity
  [OK] true=request_refund  pred=request_refund


zero_shot:  64%|██████▍   | 16/25 [28:45<15:19, 102.13s/it]

  [OK] true=activate_my_card  pred=activate_my_card
  [OK] true=getting_spare_card  pred=getting_spare_card
  [OK] true=card_about_to_expire  pred=card_about_to_expire
  [OK] true=supported_cards_and_currencies  pred=supported_cards_and_currencies
  [OK] true=cash_withdrawal_charge  pred=cash_withdrawal_charge
  [OK] true=lost_or_stolen_card  pred=lost_or_stolen_card
  [OK] true=verify_my_identity  pred=verify_my_identity
  [MISS] true=top_up_by_card_charge  pred=wrong_exchange_rate_for_cash_withdrawal


zero_shot:  68%|██████▊   | 17/25 [30:31<13:46, 103.30s/it]

  [MISS] true=compromised_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=top_up_by_cash_or_cheque  pred=top_up_by_cash_or_cheque
  [MISS] true=cash_withdrawal_not_recognised  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=card_linking  pred=card_linking
  [MISS] true=pending_transfer  pred=transfer_timing
  [OK] true=card_delivery_estimate  pred=card_delivery_estimate
  [MISS] true=getting_virtual_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=balance_not_updated_after_cheque_or_cash_deposit  pred=balance_not_updated_after_cheque_or_cash_deposit


zero_shot:  72%|███████▏  | 18/25 [32:28<12:31, 107.39s/it]

  [OK] true=top_up_by_card_charge  pred=top_up_by_card_charge
  [OK] true=terminate_account  pred=terminate_account
  [OK] true=declined_cash_withdrawal  pred=declined_cash_withdrawal
  [OK] true=card_delivery_estimate  pred=card_delivery_estimate
  [OK] true=transfer_fee_charged  pred=transfer_fee_charged
  [OK] true=card_not_working  pred=card_not_working
  [OK] true=edit_personal_details  pred=edit_personal_details
  [OK] true=transaction_charged_twice  pred=transaction_charged_twice


zero_shot:  76%|███████▌  | 19/25 [33:35<09:30, 95.04s/it] 

  [MISS] true=card_linking  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=pending_top_up  pred=pending_top_up
  [OK] true=unable_to_verify_identity  pred=unable_to_verify_identity
  [OK] true=activate_my_card  pred=activate_my_card
  [MISS] true=extra_charge_on_statement  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=card_payment_not_recognised  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=order_physical_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=verify_my_identity  pred=verify_my_identity


zero_shot:  80%|████████  | 20/25 [35:22<08:13, 98.77s/it]

  [OK] true=top_up_failed  pred=top_up_failed
  [OK] true=verify_top_up  pred=verify_top_up
  [OK] true=card_payment_fee_charged  pred=card_payment_fee_charged
  [OK] true=supported_cards_and_currencies  pred=supported_cards_and_currencies
  [OK] true=declined_card_payment  pred=declined_card_payment
  [OK] true=card_acceptance  pred=card_acceptance
  [MISS] true=age_limit  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=failed_transfer  pred=failed_transfer


zero_shot:  84%|████████▍ | 21/25 [37:25<07:03, 105.95s/it]

  [MISS] true=disposable_card_limits  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=declined_cash_withdrawal  pred=declined_cash_withdrawal
  [MISS] true=unable_to_verify_identity  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=verify_my_identity  pred=verify_my_identity
  [MISS] true=direct_debit_payment_not_recognised  pred=receiving_money
  [OK] true=age_limit  pred=age_limit
  [OK] true=atm_support  pred=atm_support
  [OK] true=card_swallowed  pred=card_swallowed


zero_shot:  88%|████████▊ | 22/25 [39:27<05:32, 110.93s/it]

  [OK] true=transfer_into_account  pred=transfer_into_account
  [OK] true=lost_or_stolen_phone  pred=lost_or_stolen_phone
  [MISS] true=automatic_top_up  pred=top_up_limits
  [OK] true=terminate_account  pred=terminate_account
  [OK] true=contactless_not_working  pred=contactless_not_working
  [OK] true=transfer_timing  pred=transfer_timing
  [MISS] true=order_physical_card  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=why_verify_identity  pred=why_verify_identity


zero_shot:  92%|█████████▏| 23/25 [41:25<03:45, 112.90s/it]

  [OK] true=verify_source_of_funds  pred=verify_source_of_funds
  [MISS] true=cash_withdrawal_not_recognised  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=card_payment_wrong_exchange_rate  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=extra_charge_on_statement  pred=extra_charge_on_statement
  [MISS] true=card_payment_fee_charged  pred=extra_charge_on_statement
  [MISS] true=top_up_by_cash_or_cheque  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=transfer_timing  pred=transfer_timing
  [OK] true=transaction_charged_twice  pred=transaction_charged_twice


zero_shot:  96%|█████████▌| 24/25 [43:27<01:55, 115.57s/it]

  [MISS] true=wrong_amount_of_cash_received  pred=declined_cash_withdrawal
  [OK] true=virtual_card_not_working  pred=virtual_card_not_working
  [MISS] true=top_up_by_bank_transfer_charge  pred=transfer_into_account
  [OK] true=card_swallowed  pred=card_swallowed
  [OK] true=card_payment_fee_charged  pred=card_payment_fee_charged
  [MISS] true=direct_debit_payment_not_recognised  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=balance_not_updated_after_cheque_or_cash_deposit  pred=balance_not_updated_after_cheque_or_cash_deposit
  [OK] true=card_swallowed  pred=card_swallowed


zero_shot: 100%|██████████| 25/25 [45:29<00:00, 109.18s/it]


>>> zero_shot accuracy: 0.6400 (128/200)

                                                  precision    recall  f1-score   support

                           Refund_not_showing_up     1.0000    1.0000    1.0000         2
                                activate_my_card     1.0000    1.0000    1.0000         3
                                       age_limit     1.0000    0.3333    0.5000         3
                         apple_pay_or_google_pay     1.0000    0.6667    0.8000         3
                                     atm_support     0.6667    1.0000    0.8000         2
                                automatic_top_up     1.0000    0.3333    0.5000         3
         balance_not_updated_after_bank_transfer     0.0000    0.0000    0.0000         2
balance_not_updated_after_cheque_or_cash_deposit     1.0000    0.6667    0.8000         3
                         beneficiary_not_allowed     0.0000    0.0000    0.0000         3
                                 cancel_transfer     1.0


c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is",

VRAM freed: 0.06GB used
[FINETUNED] Loading model: c:\Users\VINH\OneDrive - VNU-HCMUS\Attachments\Desktop\YEAR 3\ƯDNLP\lab_2\banking-intent-unsloth\outputs\checkpoint
==((====))==  Unsloth 2026.4.6: Fast Qwen3 patching. Transformers: 4.57.0.
   \\   /|    NVIDIA GeForce RTX 3070 Laptop GPU. Num GPUs = 1. Max memory: 8.0 GB. Platform: Windows.
O^O/ \_/ \    Torch: 2.7.1+cu118. CUDA: 8.6. CUDA Toolkit: 11.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Unsloth 2026.4.6 patched 36 layers with 0 QKV layers, 0 O layers and 0 MLP layers.


LangSmith tracing enabled — project: banking-intent-unsloth


finetuned:   0%|          | 0/25 [00:00<?, ?it/s]

  [OK] true=virtual_card_not_working  pred=virtual_card_not_working
  [OK] true=change_pin  pred=change_pin
  [OK] true=apple_pay_or_google_pay  pred=apple_pay_or_google_pay
  [MISS] true=top_up_by_bank_transfer_charge  pred=transfer_fee_charged
  [OK] true=automatic_top_up  pred=automatic_top_up
  [OK] true=transfer_not_received_by_recipient  pred=transfer_not_received_by_recipient
  [OK] true=beneficiary_not_allowed  pred=beneficiary_not_allowed
  [OK] true=transfer_into_account  pred=transfer_into_account


finetuned:   4%|▍         | 1/25 [00:04<01:43,  4.33s/it]

  [OK] true=receiving_money  pred=receiving_money
  [OK] true=lost_or_stolen_card  pred=lost_or_stolen_card
  [OK] true=verify_top_up  pred=verify_top_up
  [MISS] true=balance_not_updated_after_bank_transfer  pred=pending_transfer
  [OK] true=exchange_charge  pred=exchange_charge
  [OK] true=top_up_failed  pred=top_up_failed
  [OK] true=top_up_by_cash_or_cheque  pred=top_up_by_cash_or_cheque
  [OK] true=passcode_forgotten  pred=passcode_forgotten


finetuned:   8%|▊         | 2/25 [00:08<01:40,  4.39s/it]

  [MISS] true=pending_top_up  pred=top_up_reverted
  [OK] true=card_about_to_expire  pred=card_about_to_expire
  [OK] true=wrong_amount_of_cash_received  pred=wrong_amount_of_cash_received
  [MISS] true=top_up_reverted  pred=top_up_failed
  [OK] true=failed_transfer  pred=failed_transfer
  [OK] true=supported_cards_and_currencies  pred=supported_cards_and_currencies
  [OK] true=unable_to_verify_identity  pred=unable_to_verify_identity
  [OK] true=top_up_limits  pred=top_up_limits


finetuned:  12%|█▏        | 3/25 [00:12<01:31,  4.17s/it]

  [OK] true=getting_virtual_card  pred=getting_virtual_card
  [OK] true=balance_not_updated_after_bank_transfer  pred=balance_not_updated_after_bank_transfer
  [OK] true=pending_transfer  pred=pending_transfer
  [OK] true=exchange_rate  pred=exchange_rate
  [OK] true=exchange_via_app  pred=exchange_via_app
  [OK] true=declined_transfer  pred=declined_transfer
  [MISS] true=transfer_not_received_by_recipient  pred=pending_transfer
  [OK] true=cash_withdrawal_charge  pred=cash_withdrawal_charge


finetuned:  16%|█▌        | 4/25 [00:16<01:28,  4.20s/it]

  [OK] true=get_physical_card  pred=get_physical_card
  [OK] true=card_not_working  pred=card_not_working
  [MISS] true=automatic_top_up  pred=top_up_by_card_charge
  [OK] true=lost_or_stolen_card  pred=lost_or_stolen_card
  [OK] true=reverted_card_payment?  pred=reverted_card_payment?
  [OK] true=get_physical_card  pred=get_physical_card
  [OK] true=declined_cash_withdrawal  pred=declined_cash_withdrawal
  [MISS] true=pending_top_up  pred=top_up_reverted


finetuned:  20%|██        | 5/25 [00:20<01:22,  4.13s/it]

  [OK] true=edit_personal_details  pred=edit_personal_details
  [MISS] true=transfer_fee_charged  pred=card_payment_fee_charged
  [OK] true=pending_cash_withdrawal  pred=pending_cash_withdrawal
  [OK] true=contactless_not_working  pred=contactless_not_working
  [OK] true=fiat_currency_support  pred=fiat_currency_support
  [OK] true=fiat_currency_support  pred=fiat_currency_support
  [OK] true=pin_blocked  pred=pin_blocked
  [OK] true=card_payment_wrong_exchange_rate  pred=card_payment_wrong_exchange_rate


finetuned:  24%|██▍       | 6/25 [00:25<01:18,  4.15s/it]

  [OK] true=wrong_exchange_rate_for_cash_withdrawal  pred=wrong_exchange_rate_for_cash_withdrawal
  [OK] true=compromised_card  pred=compromised_card
  [OK] true=top_up_by_bank_transfer_charge  pred=top_up_by_bank_transfer_charge
  [OK] true=failed_transfer  pred=failed_transfer
  [MISS] true=getting_virtual_card  pred=get_disposable_virtual_card
  [OK] true=declined_card_payment  pred=declined_card_payment
  [MISS] true=declined_transfer  pred=declined_card_payment
  [OK] true=edit_personal_details  pred=edit_personal_details


finetuned:  28%|██▊       | 7/25 [00:29<01:15,  4.21s/it]

  [OK] true=visa_or_mastercard  pred=visa_or_mastercard
  [OK] true=get_physical_card  pred=get_physical_card
  [OK] true=card_arrival  pred=card_arrival
  [OK] true=beneficiary_not_allowed  pred=beneficiary_not_allowed
  [MISS] true=change_pin  pred=get_physical_card
  [MISS] true=topping_up_by_card  pred=top_up_by_cash_or_cheque
  [OK] true=transfer_not_received_by_recipient  pred=transfer_not_received_by_recipient
  [OK] true=apple_pay_or_google_pay  pred=apple_pay_or_google_pay


finetuned:  32%|███▏      | 8/25 [00:33<01:12,  4.27s/it]

  [OK] true=getting_spare_card  pred=getting_spare_card
  [OK] true=cancel_transfer  pred=cancel_transfer
  [OK] true=receiving_money  pred=receiving_money
  [OK] true=change_pin  pred=change_pin
  [OK] true=country_support  pred=country_support
  [OK] true=activate_my_card  pred=activate_my_card
  [OK] true=card_payment_wrong_exchange_rate  pred=card_payment_wrong_exchange_rate
  [OK] true=get_disposable_virtual_card  pred=get_disposable_virtual_card


finetuned:  36%|███▌      | 9/25 [00:37<01:07,  4.20s/it]

  [OK] true=pending_card_payment  pred=pending_card_payment
  [OK] true=transfer_fee_charged  pred=transfer_fee_charged
  [OK] true=cash_withdrawal_not_recognised  pred=cash_withdrawal_not_recognised
  [OK] true=pending_cash_withdrawal  pred=pending_cash_withdrawal
  [OK] true=transfer_into_account  pred=transfer_into_account
  [OK] true=passcode_forgotten  pred=passcode_forgotten
  [MISS] true=card_about_to_expire  pred=card_arrival
  [MISS] true=atm_support  pred=cash_withdrawal_not_recognised


finetuned:  40%|████      | 10/25 [00:42<01:02,  4.20s/it]

  [OK] true=card_acceptance  pred=card_acceptance
  [OK] true=cancel_transfer  pred=cancel_transfer
  [OK] true=card_linking  pred=card_linking
  [OK] true=wrong_exchange_rate_for_cash_withdrawal  pred=wrong_exchange_rate_for_cash_withdrawal
  [MISS] true=get_disposable_virtual_card  pred=disposable_card_limits
  [OK] true=receiving_money  pred=receiving_money
  [OK] true=card_payment_not_recognised  pred=card_payment_not_recognised
  [MISS] true=order_physical_card  pred=card_arrival


finetuned:  44%|████▍     | 11/25 [00:46<01:00,  4.31s/it]

  [OK] true=top_up_reverted  pred=top_up_reverted
  [OK] true=country_support  pred=country_support
  [MISS] true=card_delivery_estimate  pred=card_arrival
  [OK] true=exchange_charge  pred=exchange_charge
  [MISS] true=pending_card_payment  pred=pending_transfer
  [OK] true=declined_card_payment  pred=declined_card_payment
  [OK] true=passcode_forgotten  pred=passcode_forgotten
  [OK] true=beneficiary_not_allowed  pred=beneficiary_not_allowed


finetuned:  48%|████▊     | 12/25 [00:50<00:54,  4.16s/it]

  [OK] true=verify_source_of_funds  pred=verify_source_of_funds
  [MISS] true=top_up_by_card_charge  pred=top_up_by_bank_transfer_charge
  [OK] true=request_refund  pred=request_refund
  [OK] true=Refund_not_showing_up  pred=Refund_not_showing_up
  [OK] true=Refund_not_showing_up  pred=Refund_not_showing_up
  [OK] true=country_support  pred=country_support
  [OK] true=card_not_working  pred=card_not_working
  [OK] true=exchange_rate  pred=exchange_rate


finetuned:  52%|█████▏    | 13/25 [00:55<00:51,  4.32s/it]

  [OK] true=fiat_currency_support  pred=fiat_currency_support
  [OK] true=pending_card_payment  pred=pending_card_payment
  [OK] true=terminate_account  pred=terminate_account
  [OK] true=disposable_card_limits  pred=disposable_card_limits
  [MISS] true=declined_transfer  pred=declined_card_payment
  [OK] true=reverted_card_payment?  pred=reverted_card_payment?
  [OK] true=pin_blocked  pred=pin_blocked
  [OK] true=exchange_charge  pred=exchange_charge


finetuned:  56%|█████▌    | 14/25 [00:59<00:47,  4.28s/it]

  [OK] true=card_arrival  pred=card_arrival
  [OK] true=top_up_limits  pred=top_up_limits
  [OK] true=cash_withdrawal_charge  pred=cash_withdrawal_charge
  [MISS] true=balance_not_updated_after_cheque_or_cash_deposit  pred=top_up_by_cash_or_cheque
  [OK] true=visa_or_mastercard  pred=visa_or_mastercard
  [OK] true=top_up_reverted  pred=top_up_reverted
  [OK] true=card_acceptance  pred=card_acceptance
  [OK] true=apple_pay_or_google_pay  pred=apple_pay_or_google_pay


finetuned:  60%|██████    | 15/25 [01:03<00:43,  4.33s/it]

  [OK] true=wrong_amount_of_cash_received  pred=wrong_amount_of_cash_received
  [OK] true=top_up_limits  pred=top_up_limits
  [OK] true=age_limit  pred=age_limit
  [OK] true=cancel_transfer  pred=cancel_transfer
  [OK] true=pending_cash_withdrawal  pred=pending_cash_withdrawal
  [OK] true=visa_or_mastercard  pred=visa_or_mastercard
  [OK] true=why_verify_identity  pred=why_verify_identity
  [OK] true=request_refund  pred=request_refund


finetuned:  64%|██████▍   | 16/25 [01:07<00:38,  4.27s/it]

  [OK] true=activate_my_card  pred=activate_my_card
  [OK] true=getting_spare_card  pred=getting_spare_card
  [OK] true=card_about_to_expire  pred=card_about_to_expire
  [OK] true=supported_cards_and_currencies  pred=supported_cards_and_currencies
  [OK] true=cash_withdrawal_charge  pred=cash_withdrawal_charge
  [OK] true=lost_or_stolen_card  pred=lost_or_stolen_card
  [OK] true=verify_my_identity  pred=verify_my_identity
  [MISS] true=top_up_by_card_charge  pred=supported_cards_and_currencies


finetuned:  68%|██████▊   | 17/25 [01:12<00:34,  4.27s/it]

  [OK] true=compromised_card  pred=compromised_card
  [OK] true=top_up_by_cash_or_cheque  pred=top_up_by_cash_or_cheque
  [OK] true=cash_withdrawal_not_recognised  pred=cash_withdrawal_not_recognised
  [OK] true=card_linking  pred=card_linking
  [MISS] true=pending_transfer  pred=transfer_timing
  [OK] true=card_delivery_estimate  pred=card_delivery_estimate
  [OK] true=getting_virtual_card  pred=getting_virtual_card
  [OK] true=balance_not_updated_after_cheque_or_cash_deposit  pred=balance_not_updated_after_cheque_or_cash_deposit


finetuned:  72%|███████▏  | 18/25 [01:17<00:31,  4.50s/it]

  [OK] true=top_up_by_card_charge  pred=top_up_by_card_charge
  [OK] true=terminate_account  pred=terminate_account
  [OK] true=declined_cash_withdrawal  pred=declined_cash_withdrawal
  [OK] true=card_delivery_estimate  pred=card_delivery_estimate
  [OK] true=transfer_fee_charged  pred=transfer_fee_charged
  [OK] true=card_not_working  pred=card_not_working
  [OK] true=edit_personal_details  pred=edit_personal_details
  [OK] true=transaction_charged_twice  pred=transaction_charged_twice


finetuned:  76%|███████▌  | 19/25 [01:21<00:26,  4.48s/it]

  [OK] true=card_linking  pred=card_linking
  [OK] true=pending_top_up  pred=pending_top_up
  [OK] true=unable_to_verify_identity  pred=unable_to_verify_identity
  [OK] true=activate_my_card  pred=activate_my_card
  [MISS] true=extra_charge_on_statement  pred=pending_cash_withdrawal
  [OK] true=card_payment_not_recognised  pred=card_payment_not_recognised
  [OK] true=order_physical_card  pred=order_physical_card
  [OK] true=verify_my_identity  pred=verify_my_identity


finetuned:  80%|████████  | 20/25 [01:25<00:21,  4.27s/it]

  [OK] true=top_up_failed  pred=top_up_failed
  [OK] true=verify_top_up  pred=verify_top_up
  [OK] true=card_payment_fee_charged  pred=card_payment_fee_charged
  [MISS] true=supported_cards_and_currencies  pred=fiat_currency_support
  [OK] true=declined_card_payment  pred=declined_card_payment
  [OK] true=card_acceptance  pred=card_acceptance
  [OK] true=age_limit  pred=age_limit
  [OK] true=failed_transfer  pred=failed_transfer


finetuned:  84%|████████▍ | 21/25 [01:29<00:16,  4.13s/it]

  [OK] true=disposable_card_limits  pred=disposable_card_limits
  [OK] true=declined_cash_withdrawal  pred=declined_cash_withdrawal
  [OK] true=unable_to_verify_identity  pred=unable_to_verify_identity
  [OK] true=verify_my_identity  pred=verify_my_identity
  [MISS] true=direct_debit_payment_not_recognised  pred=verify_source_of_funds
  [OK] true=age_limit  pred=age_limit
  [OK] true=atm_support  pred=atm_support
  [OK] true=card_swallowed  pred=card_swallowed


finetuned:  88%|████████▊ | 22/25 [01:33<00:12,  4.06s/it]

  [OK] true=transfer_into_account  pred=transfer_into_account
  [OK] true=lost_or_stolen_phone  pred=lost_or_stolen_phone
  [OK] true=automatic_top_up  pred=automatic_top_up
  [OK] true=terminate_account  pred=terminate_account
  [OK] true=contactless_not_working  pred=contactless_not_working
  [OK] true=transfer_timing  pred=transfer_timing
  [MISS] true=order_physical_card  pred=card_arrival
  [OK] true=why_verify_identity  pred=why_verify_identity


finetuned:  92%|█████████▏| 23/25 [01:37<00:08,  4.07s/it]

  [OK] true=verify_source_of_funds  pred=verify_source_of_funds
  [OK] true=cash_withdrawal_not_recognised  pred=cash_withdrawal_not_recognised
  [OK] true=card_payment_wrong_exchange_rate  pred=card_payment_wrong_exchange_rate
  [OK] true=extra_charge_on_statement  pred=extra_charge_on_statement
  [OK] true=card_payment_fee_charged  pred=card_payment_fee_charged
  [OK] true=top_up_by_cash_or_cheque  pred=top_up_by_cash_or_cheque
  [OK] true=transfer_timing  pred=transfer_timing
  [OK] true=transaction_charged_twice  pred=transaction_charged_twice


finetuned:  96%|█████████▌| 24/25 [01:41<00:04,  4.16s/it]

  [MISS] true=wrong_amount_of_cash_received  pred=declined_cash_withdrawal
  [OK] true=virtual_card_not_working  pred=virtual_card_not_working
  [MISS] true=top_up_by_bank_transfer_charge  pred=receiving_money
  [OK] true=card_swallowed  pred=card_swallowed
  [OK] true=card_payment_fee_charged  pred=card_payment_fee_charged
  [OK] true=direct_debit_payment_not_recognised  pred=direct_debit_payment_not_recognised
  [OK] true=balance_not_updated_after_cheque_or_cash_deposit  pred=balance_not_updated_after_cheque_or_cash_deposit
  [OK] true=card_swallowed  pred=card_swallowed


finetuned: 100%|██████████| 25/25 [01:46<00:00,  4.27s/it]


>>> finetuned accuracy: 0.8550 (171/200)

                                                  precision    recall  f1-score   support

                           Refund_not_showing_up     1.0000    1.0000    1.0000         2
                                activate_my_card     1.0000    1.0000    1.0000         3
                                       age_limit     1.0000    1.0000    1.0000         3
                         apple_pay_or_google_pay     1.0000    1.0000    1.0000         3
                                     atm_support     1.0000    0.5000    0.6667         2
                                automatic_top_up     1.0000    0.6667    0.8000         3
         balance_not_updated_after_bank_transfer     1.0000    0.5000    0.6667         2
balance_not_updated_after_cheque_or_cash_deposit     1.0000    0.6667    0.8000         3
                         beneficiary_not_allowed     1.0000    1.0000    1.0000         3
                                 cancel_transfer     1.0


c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\VINH\miniconda3\envs\manga_env\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is",

VRAM freed: 0.06GB used
=== SUMMARY ===
  zero_shot    0.6400  (64.00%)
  finetuned    0.8550  (85.50%)


## 6. Evaluate on full test set

Run zero_shot and finetuned on all 3,080 samples.

In [ ]:
print("\n" + "=" * 60)
print(f"FULL EVALUATION ({len(test_df)} samples)")
print("=" * 60)

full_results = {}
for mode in ("zero_shot", "finetuned"):
    clf = IntentClassification(f"{WORKING_DIR}/configs/inference.yaml", mode=mode)
print("=== FULL EVALUATION SUMMARY ===")
for mode, acc in full_results.items():
    print(f"  {mode:<12} {acc:.4f}  ({acc*100:.2f}%)")



FULL EVALUATION (3080 samples)
[ZERO_SHOT] Loading model: unsloth/Qwen3-4B-unsloth-bnb-4bit
==((====))==  Unsloth 2026.4.6: Fast Qwen3 patching. Transformers: 4.57.0.
   \\   /|    NVIDIA GeForce RTX 3070 Laptop GPU. Num GPUs = 1. Max memory: 8.0 GB. Platform: Windows.
O^O/ \_/ \    Torch: 2.7.1+cu118. CUDA: 8.6. CUDA Toolkit: 11.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
unsloth/Qwen3-4B-unsloth-bnb-4bit does not have a padding token! Will use pad_token = <|PAD_TOKEN|>.
LangSmith tracing enabled — project: banking-intent-unsloth


zero_shot:  41%|████      | 157/385 [3:51:28<6:21:45, 100.46s/it] 